Generic signature API

bwesterb opened the pull request:
Used to easily add support for new (PQ) signature schemes to [our fork of Go](https://github.com/cloudflare/go) and cfssl.

bwesterb commented:
Another reason is that I'm quickly prototyping integration into Go stdlib to find possible issues with the API.  I found one already: we can't use `PubicKeyAlgorithm() x509.PublicKeyAlgorithm` in our interface for that would cause an import cycle.

bwesterb commented:
@armfazh @claucece Have another look.

claucece commented:
I'll check this tomorrow ;)

armfazh commented on sign/api.go:
```diff
@@ -0,0 +1,413 @@
+
+func UnmarshalPEMPublicKey(data []byte) (PublicKey, error) {
+	block, rest := pem.Decode(data)
+	if len(rest) != 0 {
```
I think the checks are for the block, like in the [Decode example](https://golang.org/pkg/encoding/pem/#Decode)

```
        block, rest := pem.Decode(pubPEMData)
	if block == nil || block.Type != "PUBLIC KEY" {
		log.Fatal("failed to decode PEM block containing public key")
	}
```

armfazh commented on sign/eddilithium3.go:
```diff
@@ -0,0 +1,108 @@
+
+type edDilithium3Scheme struct{}
+
+var EdDilithium3 Scheme = &edDilithium3Scheme{}
```
it does not require the address

bwesterb replied:
Try it without the `&`.

armfazh commented on sign/eddilithium3.go:
```diff
@@ -0,0 +1,108 @@
+	"errors"
+)
+
+type edDilithium3Scheme struct{}
```
All these definitions must live inside of `sign/eddilithium3` package,

bwesterb replied:
That is what I would like as well, but how do you do that without creating an import cycle?

armfazh commented on sign/eddilithium3.go:
```diff
@@ -0,0 +1,108 @@
+	"errors"
+)
+
+type edDilithium3Scheme struct{}
```
This is a zero-sized data, which is ok. What if this is an integer, which can also be used as identifier in the list of schemes.

bwesterb replied:
You can't use a specific integer as a type.

claucece commented:
I think it is looking good. Maybe we should be more clear on some names and all, though

bwesterb commented on sign/api.go:
```diff
@@ -0,0 +1,413 @@
+	Name() string
+
+	// Size of signatures
+	SignatureSize() uint
```
Sizes are always positive.  However, it might indeed be better to use `int` as it removed the need for casting, say, `len(something)` to `uint`.

claucece replied:
Yeah.. I was thinking from the casting perspective that it is used many times...

bwesterb replied:
Ok, let's go for int!

bwesterb commented on sign/api.go:
```diff
@@ -0,0 +1,413 @@
+// A Scheme represents a specific instance of a signature scheme.
+type Scheme interface {
+	// GenerateKey creates a new key-pair.
+	GenerateKey() (PublicKey, PrivateKey, error)
```
The common case should be as easy and foolproof as possible and the common case is that people don't  need to use a custom random source.

claucece replied:
I agree now.

bwesterb commented on sign/api.go:
```diff
@@ -0,0 +1,413 @@
+	// XXX add a (compile time?) lookup table
+	name = strings.ToLower(name)
+	for _, scheme := range schemes {
+		if strings.ToLower(scheme.Name()) == name {
```
If we would remove this `ToLower()` then we will never match any scheme that has an upper case character in the return of its `scheme.Name()`.

claucece replied:
Yes, but perhaps the `scheme.Name()` should always be enforced to be lower case... what do you think?

bwesterb replied:
@armfazh Removed case insensitivity again. Did you do that on purpose @armfazh ?

bwesterb commented on sign/api.go:
```diff
@@ -0,0 +1,413 @@
+// Go1.15 adds {PublicKey,PrivateKey}.Equal(). Until then, we can use that
+// we use this.
+
+func PublicKeysEqual(a, b PublicKey) bool {
```
I could use the interface that we expect Go1.15 to have, but I'd prefer to wait in case they change their mind.

bwesterb commented on sign/api.go:
```diff
@@ -0,0 +1,413 @@
+}
+
+// ListSchemeNames returns the names of all schemes supported.
+func ListSchemeNames() []string {
```
I'm thinking of adding such a function as well to return `[]Scheme`.

bwesterb replied:
`api.AllSchemes()`

bwesterb commented on sign/eddilithium3.go:
```diff
@@ -0,0 +1,108 @@
+func (s *edDilithium3Scheme) Sign(sk PrivateKey, message []byte,
+	opts *SignatureOpts) []byte {
+	sig := make([]byte, eddilithium3.SignatureSize)
+	if opts != nil && opts.Context != "" {
```
I don't understand.  For `opts.Context` to make sense, we must have `opts != nil`.

armfazh replied:
if any of the options is not used, then is ignored.

bwesterb replied:
I don't understand — could you rephrase perhaps?

bwesterb commented:
> Maybe we should be more clear on some names and all, though

Sure, which?

armfazh commented:
I just added a generic interface for four signatures, there are some methods marked with TODO.
Also I moved some methods into the `circl/pki` package.

claucece commented:
> Sure, which?

Let me thinking of a nice naming convention and propose something over here tomorrow.

bwesterb commented on sign/ed25519/signapi.go:
```diff
@@ -0,0 +1,75 @@
+
+type scheme sign.SchemeID
+
+func (scheme) ID() sign.SchemeID    { return sign.SchemeID(Scheme) }
```
What is the advantage of this compared to:

```
type scheme struct{}
func (scheme) ID() sign.SchemeID    { return sign.Ed25519 }
```

bwesterb commented on sign/api/api.go:
```diff
@@ -0,0 +1,33 @@
+	register(ed25519.Scheme)
+	register(ed448.Scheme)
+	register(eddilithium3.Scheme)
+	register(eddilithium4.Scheme)
```
Why not `register(sign.Ed25519, ed25519.Scheme)`?  Then there is no need for a public `ID()`, `SchemeID`, etc.

bwesterb replied:
@armfazh I removed `SchemeID` without changing the rest of the API at all.

bwesterb commented on sign/api/api.go:
```diff
@@ -0,0 +1,33 @@
+func SchemeByName(name string) sign.Scheme { return allSchemeNames[name] }
+
+// AllSchemes returns all signature schemes supported.
+func AllSchemes() []sign.Scheme { a := allSchemes; return a[:] }
```
This allows the caller to manipulate the list. That's dangerous.

armfazh replied:
`a` is a copy, not a reference to the register

bwesterb replied:
Yes, you're right. I didn't read it correctly.

bwesterb commented:
I have a few issues with your changes @armfazh 

- I dislike that `SchemeByName` is now under the `github.com/cloudflare/circl/sign/api` package.
- Providing a context to a scheme that does not support it is completely ignored.  It should be an error.

bwesterb commented on pki/pki.go:
```diff
@@ -0,0 +1,179 @@
+
+func init() {
+	allSchemesByOID = make(map[string]sign.Scheme)
+	for _, scheme := range api.AllSchemes() {
```
@armfazh You can't rely on the right `init()` order.

bwesterb replied:
https://medium.com/@markbates/go-init-order-dafa89fcef22

mmcloughlin replied:
> If a package has imports, the imported packages are initialized before initializing the package itself.

https://golang.org/ref/spec#Package_initialization

mmcloughlin commented on pki/pki_test.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+func TestPEM(t *testing.T) {
+	for _, scheme := range api.AllSchemes() {
```
This seems like an appropriate use for subtests.

```go
func TestPEM(t *testing.T) {
	for _, scheme := range api.AllSchemes() {
		t.Run(scheme.Name(), func(t *testing.T) {
			// ...
		}
	}
}
```

bwesterb replied:
Nice, didn't know about these.  Thanks!

mmcloughlin commented on sign/api/api_test.go:
```diff
@@ -0,0 +1,94 @@
+
+func TestApi(t *testing.T) {
+	allSchemes := api.AllSchemes()
+	for _, scheme := range allSchemes {
```
Consider subtests?

mmcloughlin commented on sign/ed25519/ed25519_test.go:
```diff
@@ -39,6 +39,24 @@ func TestMalleability(t *testing.T) {
+	if !pub.Equal(pub) {
+		t.Fatal()
+	}
+	if err != nil {
```
Why is the error check not the first assertion?

bwesterb replied:
That's a bug. Thanks for pointing that out.

mmcloughlin commented on sign/ed25519/ed25519_test.go:
```diff
@@ -39,6 +39,24 @@ func TestMalleability(t *testing.T) {
+		t.Fatal()
+	}
+	if !pub.Equal(pub) {
+		t.Fatal()
```
Why no message with `Fatal`? If you intend to use no message, just use `FailNow()` directly.

https://golang.org/pkg/testing/#T.Fatal

bwesterb replied:
Does it really matter?  (Changed anyway.)

bwesterb replied:
It does matter: FailNow doesn't show the line number when triggered and Fatal does.  Thus I reverted back to using Fatal.

mmcloughlin commented on sign/ed25519/signapi.go:
```diff
@@ -0,0 +1,75 @@
+package ed25519
```
Nearly identical to the other `signapi.go` files (I know there are some differences). Is it generated? If so it needs a generated comment.

bwesterb replied:
No, it's not.

mmcloughlin commented on sign/ed25519/signapi.go:
```diff
@@ -0,0 +1,75 @@
+func (scheme) Sign(
+	sk sign.PrivateKey,
+	message []byte,
+	opts *sign.SignatureOpts) []byte {
```
nit: my preference would be to use a trailing comma in multi-line signatures.

bwesterb replied:
Yes, that's nicer.

mmcloughlin commented on sign/sign.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+var (
+	// ErrType is
```
FYI, if you don't want to document these individually, I _think_ the linter accepts just a comment on the entire `var` block.

mmcloughlin commented on sign/sign.go:
```diff
@@ -0,0 +1,107 @@
+
+var (
+	// ErrType is
+	ErrType = errors.New("types mismatch")
```
nit: `ErrTypeMismatch`?

bwesterb commented:
@armfazh Will you make changes according to the comments or shall I make them?

armfazh commented:
> @armfazh Will you make changes according to the comments or shall I make them?

Please, go forward with changes.

bwesterb commented:
@armfazh 

> * I dislike that `SchemeByName` is now under the `github.com/cloudflare/circl/sign/api` package.

I don't see a way to solve this without making something else ugly.

> * Providing a context to a scheme that does not support it is completely ignored.  It should be an error.

I've changed that behaviour: provided a context where none is supported causes an error.

bwesterb commented:
I've also removed the `Hash` from the SignatureOpts as it's not used.

bwesterb commented on sign/api.go:
```diff
@@ -0,0 +1,413 @@
+}
+
+func UnmarshalPKIXPrivateKey(data []byte) (PrivateKey, error) {
+	var pkix struct {
```
For the private keys it is (and so I made a common type) — for the public keys it isn't.

bwesterb commented:
@armfazh @claucece I've addressed all comments.  Before starting with final polish (i.e. better documentation & examples), I think we still need to discuss some API issues:

- [ ] Are we really happy with having `sign` package which defines `Scheme` and `sign/api` package which actually lists them?  Isn't `schemes` perhaps a better name than `api` as the `sign` package itself defines the API.  Then `schemes.All()` and `scheme.ByName()` would be nice to write.
- [ ] Do we want Scheme names to be case insensitive?  This was the case but @armfazh reversed it. This is required for easy integration with `cfssl`.
- [ ] Doesn't ed25519 have a variant that support contexts?  Should we add this to the Ed25519 scheme or have a separate Ed25519Ctx scheme?  I guess it's best to have separate schemes for these variants — @claucece what do you think?

armfazh commented:
> * [ ]  Are we really happy with having `sign` package which defines `Scheme` and `sign/api` package which actually lists them?  Isn't `schemes` perhaps a better name than `api` as the `sign` package itself defines the API.  Then `schemes.All()` and `scheme.ByName()` would be nice to write.

Agree on `s/api/schemes/` 

> * [ ]  Do we want Scheme names to be case insensitive?  This was the case but @armfazh reversed it. This is required for easy integration with `cfssl`.

I don't have a strong opinion on this. Feel free to bring back that code.

> * [ ]  Doesn't ed25519 have a variant that support contexts?  Should we add this to the Ed25519 scheme or have a separate Ed25519Ctx scheme?  I guess it's best to have separate schemes for these variants — @claucece what do you think?

That variant would be another named scheme `Ed25519Ctx`, which is separated from the current `Ed25519` scheme.

> I've also removed the Hash from the SignatureOpts as it's not used — we can add it back in if we have a scheme that actually uses it.

I added that the Hash field to make the structure compatible with the crypto.SignerOpts. So, we can receive a crypto.SignerOpts, and internally convert it to a different type.

> I dislike that SchemeByName is now under the github.com/cloudflare/circl/sign/api package.

I also didn't like it at all. But, that's the way I found for avoiding circle dependencies. Renaming sign/api to sign/schemes can alleviate a bit its usage.

armfazh commented on sign/ed25519/signapi.go:
```diff
@@ -0,0 +1,84 @@
+	"github.com/cloudflare/circl/sign"
+)
+
+var Scheme sign.Scheme = &scheme{}
```
I don't understand why to take the address of a zero-sized value. You can implement the interface with `scheme` and not with `*scheme`.

bwesterb replied:
Under the hood an instance of an interface is always a pair of a pointer and a concrete type, so instantiating it with a pointer is closer to what really happens.  The assembly generated btw is exactly [the same](https://godbolt.org/z/6zq31x).

armfazh replied:
The difference is on the assembly calls.
     `call  AX` vs `call    "".(*Impl2).A(SB)`
one call is with a register and the other is with a pointer. 

I am not convinced why to use pointer when they are not needed.

bwesterb replied:
Oops, I simplified my example too much.

[Take a look at this one](https://godbolt.org/z/KsaccE)

Creating an instance of an interface with either a empty struct or a pointer to an empty struct results in the pointer `runtime.zerobase`.  (The code of `alloc1` and `alloc2` is the same.  This time I checked properly.)

The generic way to run a function on an interface involves a virtual call (`call ax`).  So interestingly the Go compiler optimised the call with the pointer receiver, but not the call with the value receiver.  I don't see a reason why it couldn't optimise the one with the value receiver as well and perhaps it will in the future.

So using pointers or not does not actually change anything when using empty structs (and the calling code doesn't know the types.) And if it does it optimizes the pointer case better. That's not the reason I use pointers though — as I said I prefer putting them here as it's closer to what Go actually does under the hood: an instance of an interface is always a pointer.

armfazh replied:
> So interestingly the Go compiler optimised the call with the pointer receiver, but not the call with the value receiver.

I want to learn more about this.

> So using pointers or not does not actually change anything when using empty structs.

Agree.

armfazh commented on sign/sign.go:
```diff
@@ -0,0 +1,103 @@
+	SeedSize() int
+
+	// Returns whether contexts are supported
+	SupportsContext() bool
```
I am not sure whether this method is required. As per the policy mentioned above, if context is provided and not required then it should fail.

bwesterb replied:
It's nice to know beforehand whether a scheme will by useable when you require a context.  (I need this in the tests to check whether to provide a context or not.)

armfazh replied:
Is this the only use case? I think this issue was one of the reasons of the previous policy. A scheme uses the context if the scheme supports it, otherwise it is simply ignored. 
Anyway, either policy works.

bwesterb replied:
Ignoring context if the scheme doesn't support it is very dangerous: a protocol might essentially require different signatures for different contexts.  If we ignore the context for schemes that don't support it, we leave a foot gun.

armfazh replied:
good point.

bwesterb commented:
> Agree on `s/api/schemes/`

Done.  Updated this branch, Go fork and cfssl fork.

> I added that the Hash field to make the structure compatible with the crypto.SignerOpts. So, we can receive a crypto.SignerOpts, and internally convert it to a different type.

The type of `Scheme.Sign` is `Sign(sk PrivateKey, message []byte, opts *SignatureOpts) []byte` so that wouldn't accept a `crypto.SignerOpts` anyway now.  (Note that there are two sign functions: one on the scheme and one on the PrivateKey — the one on the private key does accept a `crypto.SignerOpts` but does not allow for the advanced options like `Context`.)

armfazh commented:
> the one on the private key does accept a crypto.SignerOpts but does not allow for the advanced options like Context.)

I had the same intuition, but here is the trick. Even if we announce to receive a `crypto.SignerOpts`, we could pass a `Sign.SignatureOpts` (which must have field Hash) and internally convert the parameter into a `Sign.SignatureOpts` struct. (I learn this trick from @claucece)

claucece commented:
>  Are we really happy with having sign package which defines Scheme and sign/api package which actually lists them? Isn't schemes perhaps a better name than api as the sign package itself defines the API. Then schemes.All() and scheme.ByName() would be nice to write.

Yes! 

> Do we want Scheme names to be case insensitive? This was the case but @armfazh reversed it. This is required for easy integration with cfssl.

Mmm... It will be better case insensitive.

> Doesn't ed25519 have a variant that support contexts? Should we add this to the Ed25519 scheme or have a separate Ed25519Ctx scheme? I guess it's best to have separate schemes for these variants — @claucece what do you think?

There are two: the ctx and the ph. The schemes should be there on the ed25519 API.

claucece commented:
> I had the same intuition, but here is the trick. Even if we announce to receive a crypto.SignerOpts, we could pass a Sign.SignatureOpts (which must have field Hash) and internally convert the parameter into a Sign.SignatureOpts struct. (I learn this trick from @claucece)

Exactly! Then, the "Golang-way" API does not change ;)

claucece commented:
@bwesterb check what I did with the ed448 and ed25519 API, specially the usage of this: https://github.com/cloudflare/circl/blob/master/sign/ed25519/ed25519.go#L69

bwesterb commented:
> I had the same intuition, but here is the trick. Even if we announce to receive a `crypto.SignerOpts`, we could pass a `Sign.SignatureOpts` (which must have field Hash) and internally convert the parameter into a `Sign.SignatureOpts` struct. (I learn this trick from @claucece)

Yes, that's for a method on the private key, but here we're dealing with a top level function.

(For the `Sign` method on the private key it's actually a bit tricky.  If we use the same type for an ed25519 and ed25519ctx private key, then we can't know whether it was used for the scheme ed25519 or ed25519ctx.  I'd like to propose to fix this in a coming pull request where we add schemes for the variants of ed*.)

bwesterb commented:
@armfazh @claucece I'm happy with the PR as it is now (assuming we'll be able to make changes to the API before we tag it.) Do you agree?

claucece commented:
> (For the Sign method on the private key it's actually a bit tricky. If we use the same type for an ed25519 and ed25519ctx private key, then we can't know whether it was used for the scheme ed25519 or ed25519ctx. I'd like to propose to fix this in a coming pull request where we add schemes for the variants of ed*.)

Mmm.. ok.

> Do you agree?

Sure!

armfazh commented on pki/pki.go:
```diff
@@ -0,0 +1,184 @@
+		return nil, err
+	}
+	str := pem.EncodeToMemory(&pem.Block{
+		Type:  fmt.Sprintf("%s PRIVATE KEY", sk.Scheme().Name()),
```
```suggestion
		Type:  sk.Scheme().Name() + " PRIVATE KEY",
```

armfazh replied:
During unmarshalling, does this name is verified that matches with the content?

bwesterb replied:
> During unmarshalling, does this name is verified that matches with the content?

No.  The reason is that I'm not sure that all private key files actually contain the type in them.

armfazh commented on pki/pki.go:
```diff
@@ -0,0 +1,184 @@
+		},
+		asn1.BitString{
+			Bytes:     data,
+			BitLength: len(data) * 8,
```
Do we need to verify that`scheme.PublicKeySize() == len(data)` ?

bwesterb replied:
I don't think it's necessary here.

armfazh commented on pki/pki.go:
```diff
@@ -0,0 +1,184 @@
+	}
+
+	allSchemesByTLS = make(map[uint]sign.Scheme)
+	for _, scheme := range schemes.All() {
```
Merge for-loops.

bwesterb commented:
Thank you, I merged it.  I created an issue ( #147 ) to track desired improvements to this API.
